### ISSUE TITLE
Recursive types not matching properly

### ISSUE THREAD

**Issue by mcmanustfj** [microsoft/pyright#2693]
**Describe the bug**
I believe the following code should pass Pylance/pyright type checking

```python
import typing
from typing import Union, Tuple

JSONValue = Union[
    dict[str, "JSONValue"],
    list["JSONValue"],
    str,
    int,
    float,
    bool,
    None,
]
JSON = Union[dict[str, JSONValue], list[JSONValue]]


def example() -> JSON:
    partial_result: list[dict[str, str]] = [{}]
    return partial_result
```

This gets the error (over the `return partial_result` line) 
```
Expression of type "tuple[list[dict[str, str]], Literal[200]]" cannot be assigned to return type "JSONResponse"
  Tuple entry 1 is incorrect type
    Type "list[dict[str, str]]" cannot be assigned to type "JSON"
      "list[dict[str, str]]" is incompatible with "dict[str, JSONValue]"
      TypeVar "_T@list" is invariant
        Type "dict[str, str]" cannot be assigned to type "JSONValue"
          TypeVar "_VT@dict" is invariantPylancereportGeneralTypeIssues
```
even though `list[dict[str, str]]` can clearly be resolved to `list[dict[str, JSONValue]] ->   list[JSONValue] -> JSON`

**VS Code extension or command-line**
Pylance v2021.12.1

**Comment by erictraut** [comment]
Pylance is correct in flagging this as an error. The problem is that `list` is mutable container, so its type parameter is invariant. For example, this produces an error:

```python
a = [1, 2, 3]
b: list[float] = a # Error: list[int] cannot be assigned to declared type list[float]
```

In your case, you are attempting to assign a `list[dict[str, str]]` to a `list[JSONValue]`.

If you change your code to the following, it will work without error:

```python
def example() -> JSON:
    partial_result: list[JSONValue] = [{}]
    return partial_result
```
